cmd/juju/cloud/add_test.go

```diff
@@ -565,3 +650,96 @@ func (*addSuite) TestSpecifyingCloudFileThroughFlagAndArgument_Errors(c *gc.C) {
 	_, err := cmdtesting.RunCommand(c, command, "garage-maas", "-f", "fake.yaml", "foo.yaml")
 	c.Check(err, gc.ErrorMatches, "cannot specify cloud file with flag and argument")
 }
+
+func (*addSuite) TestValidateGoodCloudFile(c *gc.C) {
+	data := `
+clouds:
+  foundations:
+    type: maas
+    auth-types: [oauth1]
+    endpoint: "http://10.245.31.100/MAAS"`
+
+	cloudFile := prepareTestCloudYaml(c, data)
+	defer cloudFile.Close()
+	defer os.Remove(cloudFile.Name())
+
+	var logWriter loggo.TestWriter
+	writerName := "add_cloud_tests_writer"
+	c.Assert(loggo.RegisterWriter(writerName, &logWriter), jc.ErrorIsNil)
+	defer func() {
+		loggo.RemoveWriter(writerName)
+		logWriter.Clear()
+	}()
+
+	mockCloud, err := cloudfile.ParseCloudMetadataFile(cloudFile.Name())
+	c.Assert(err, jc.ErrorIsNil)
+
+	fake := newFakeCloudMetadataStore()
+	fake.Call("ParseCloudMetadataFile", cloudFile.Name()).Returns(mockCloud, nil)
+	fake.Call("PersonalCloudMetadata").Returns(map[string]cloudfile.Cloud{}, nil)
+	fake.Call("PublicCloudMetadata", []string(nil)).Returns(map[string]cloudfile.Cloud{}, false, nil)
+	fake.Call("WritePersonalCloudMetadata", mockCloud).Returns(nil)
+
+	command := cloud.NewAddCloudCommand(fake)
+
+	_, err = cmdtesting.RunCommand(c, command, "foundations", cloudFile.Name())
+	c.Check(err, jc.ErrorIsNil)
+
+	c.Check(logWriter.Log(), jc.LogMatches, []jc.SimpleMessage{})
+}
+
+func (*addSuite) TestValidateBadCloudFile(c *gc.C) {
+	data := `
+clouds:
+  foundations:
+    type: maas
+    auth-typs: [oauth1]
+    endpoint: "http://10.245.31.100/MAAS"`
+
+	cloudFile := prepareTestCloudYaml(c, data)
+	defer cloudFile.Close()
+	defer os.Remove(cloudFile.Name())
+
+	mockCloud, err := cloudfile.ParseCloudMetadataFile(cloudFile.Name())
+	c.Assert(err, jc.ErrorIsNil)
+
+	fake := newFakeCloudMetadataStore()
+	fake.Call("ParseCloudMetadataFile", cloudFile.Name()).Returns(mockCloud, nil)
+	fake.Call("PersonalCloudMetadata").Returns(map[string]cloudfile.Cloud{}, nil)
+	fake.Call("PublicCloudMetadata", []string(nil)).Returns(map[string]cloudfile.Cloud{}, false, nil)
+	fake.Call("WritePersonalCloudMetadata", mockCloud).Returns(nil)
+
+	command := cloud.NewAddCloudCommand(fake)
+
+	var logWriter loggo.TestWriter
+	writerName := "add_cloud_tests_writer"
+	c.Assert(loggo.RegisterWriter(writerName, &logWriter), jc.ErrorIsNil)
+	defer func() {
+		loggo.RemoveWriter(writerName)
+		logWriter.Clear()
+	}()
+
+	_, err = cmdtesting.RunCommand(c, command, "foundations", cloudFile.Name())
+	c.Check(err, jc.ErrorIsNil)
+
+	c.Check(logWriter.Log(), jc.LogMatches, []jc.SimpleMessage{
+		jc.SimpleMessage{
+			Level:   loggo.WARNING,
+			Message: `property "auth-typs" is invalid. Perhaps you mean "auth-types".`,
+		},
+	})
+}
```

Comment:
Yes, a good example would be if the user mistakenly uses the property `typ` when then were meaning to use `type`. This would lead to the following

```
WARNING
The property `typ` is invalid ...
ERROR the provider type *blank*  is not a registered ...
```
We get this error since the `type` property was misspelled. It is a reasonable error, yet probably not the best from a user's perspective.